cli/firmware/getversion.go

```diff
@@ -0,0 +1,135 @@
+	License as published by the Free Software Foundation; either
+	version 2.1 of the License, or (at your option) any later version.
+
+	This library is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+	Lesser General Public License for more details.
+
+	You should have received a copy of the GNU Lesser General Public
+	License along with this library; if not, write to the Free Software
+	Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+*/
+
+package firmware
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/arduino/arduino-cli/cli/errorcodes"
+	"github.com/arduino/arduino-cli/cli/feedback"
+	"github.com/arduino/arduino-fwuploader/cli/common"
+	"github.com/arduino/arduino-fwuploader/flasher"
+	"github.com/arduino/arduino-fwuploader/indexes/download"
+	"github.com/arduino/arduino-fwuploader/indexes/firmwareindex"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	semver "go.bug.st/relaxed-semver"
+)
+
+// NewGetVersionCommand creates a new `get-version` command
+func NewGetVersionCommand() *cobra.Command {
+
+	command := &cobra.Command{
+		Use:   "get-version",
+		Short: "Gets the version of the firmware the board is using.",
+		Long:  "Flashes a sketch to a board to obtain the firmware version used by the board",
+		Example: "" +
+			"  " + os.Args[0] + " firmware get-version --fqbn arduino:samd:mkr1000 --address COM10\n" +
+			"  " + os.Args[0] + " firmware get-version -b arduino:samd:mkr1000 -a COM10\n",
+		Args: cobra.NoArgs,
+		Run:  runGetVersion,
+	}
+	commonFlags.AddToCommand(command)
+	return command
+}
+
+func runGetVersion(cmd *cobra.Command, args []string) {
+
+	packageIndex, firmwareIndex := common.InitIndexes()
+	common.CheckFlags(commonFlags.Fqbn, commonFlags.Address)
+	board := common.GetBoard(firmwareIndex, commonFlags.Fqbn)
+	uploadToolDir := common.GetUploadToolDir(packageIndex, board)
+
+	versionSketchPath, err := download.DownloadSketch(board.VersionSketch)
+	if err != nil {
+		feedback.Errorf("Error downloading loader sketch from %s: %s", board.LoaderSketch.URL, err)
+		os.Exit(errorcodes.ErrGeneric)
+	}
+	logrus.Debugf("version sketch downloaded in %s", versionSketchPath.String())
+
+	versionSketch := strings.ReplaceAll(versionSketchPath.String(), versionSketchPath.Ext(), "")
+
+	programmerOut, programmerErr, err := common.FlashSketch(board, versionSketch, uploadToolDir, commonFlags.Address)
+	if err != nil {
+		feedback.Error(err)
+		os.Exit(errorcodes.ErrGeneric)
+	}
+
+	// Wait a bit after flashing the sketch for the board to become available again.
+	logrus.Debug("sleeping for 3 sec")
+	time.Sleep(3 * time.Second)
+
+	currentVersion, err := getVersion(board)
+	if err != nil {
+		feedback.Error(err)
+		os.Exit(1)
+	}
+	feedback.Printf("Firmware version installed: %s", currentVersion)
```

Comment:
The JSON output can't be parsed:
```json
$ ./arduino-fwuploader.exe firmware get-version -a COM14 --fqbn arduino:samd:mkrwifi1010 --format json
"Firmware version installed: 1.4.8\r"
{
  "programmer": {
    "stdout": "Set binary mode\r\nreadWord(addr=0)=0x20007ffc\r\nreadWord(addr=0xe000ed00)=0x410cc601\r\nreadWord(addr=0x41002018)=0x10010305\r\nversion()=v2.0 [Arduino:XYZ] Apr 11 2019 13:09:56\r\nchipId=0x10010005\r\nConnected at 921600 baud\r\nreadWord(addr=0)=0x20007ffc\r\nreadWord(addr=0xe000ed00)=0x410cc601\r\nreadWord(addr=0x41002018)=0x10010305\r\nAtmel SMART device 0x10010005 found\r\nwrite(addr=0x20004000,size=0x34)\r\nwriteWord(addr=0x20004030,value=0x10)\r\nwriteWord(addr=0x20004020,value=0x20008000)\r\nDevice       : ATSAMD21G18A\r\nreadWord(addr=0)=0x20007ffc\r\nreadWord(addr=0xe000ed00)=0x410cc601\r\nreadWord(addr=0x41002018)=0x10010305\r\nChip ID      : 10010005\r\nversion()=v2.0 [Arduino:XYZ] Apr 11 2019 13:09:56\r\nVersion      : v2.0 [Arduino:XYZ] Apr 11 2019 13:09:56\r\nAddress      : 8192\r\nPages        : 3968\r\nPage Size    : 64 bytes\r\nTotal Size   : 248KB\r\nPlanes       : 1\r\nLock Regions : 16\r\nLocked       : readWord(addr=0x41004020)=0xffff\r\nreadWord(addr=0x41004020)=0xffff\r\nreadWord(addr=0x41004020)=0xffff\r\nreadWord(addr=0x41004020)=0xffff\r\nreadWord(addr=0x41004020)=0xffff\r\nreadWord(addr=0x41004020)=0xffff\r\nreadWord(addr=0x41004020)=0xffff\r\nreadWord(addr=0x41004020)=0xffff\r\nreadWord(addr=0x41004020)=0xffff\r\nreadWord(addr=0x41004020)=0xffff\r\nreadWord(addr=0x41004020)=0xffff\r\nreadWord(addr=0x41004020)=0xffff\r\nreadWord(addr=0x41004020)=0xffff\r\nreadWord(addr=0x41004020)=0xffff\r\nreadWord(addr=0x41004020)=0xffff\r\nreadWord(addr=0x41004020)=0xffff\r\nnone\r\nreadWord(addr=0x41004018)=0\r\nSecurity     : false\r\nBoot Flash   : true\r\nreadWord(addr=0x40000834)=0\r\nBOD          : false\r\nreadWord(addr=0x40000834)=0\r\nBOR          : false\r\nArduino      : FAST_CHIP_ERASE\r\nArduino      : FAST_MULTI_PAGE_WRITE\r\nArduino      : CAN_CHECKSUM_MEMORY_BUFFER\r\nErase flash\r\nchipErase(addr=0x2000)\r\ndone in 0.841 seconds\r\n\r\nWrite 15588 bytes to flash (244 pages)\r\nwrite(addr=0x20005000,size=0x1000)\r\nwriteBuffer(scr_addr=0x20005000, dst_addr=0x2000, size=0x1000)\r\n\r[=======                       ] 26% 
(64/244 pages)write(addr=0x20005000,size=0x1000)\r\nwriteBuffer(scr_addr=0x20005000, dst_addr=0x3000, size=0x1000)\r\n\r[===============               ] 52% (128/244 pages)write(addr=0x20005000,size=0x1000)\r\nwriteBuffer(scr_addr=0x20005000, dst_addr=0x4000, size=0x1000)\r\n\r[=======================       ] 78% (192/244 pages)write(addr=0x20005000,size=0xd00)\r\nwriteBuffer(scr_addr=0x20005000, dst_addr=0x5000, size=0xd00)\r\n\r[==============================] 100% (244/244 pages)\r\ndone in 0.117 seconds\r\n\r\nVerify 15588 bytes of flash with checksum.\r\nchecksumBuffer(start_addr=0x2000, size=0x1000) = e84f\r\nchecksumBuffer(start_addr=0x3000, size=0x1000) = e256\r\nchecksumBuffer(start_addr=0x4000, size=0x1000) = b123\r\nchecksumBuffer(start_addr=0x5000, size=0xce4) = cfbc\r\nVerify successful\r\ndone in 0.016 seconds\r\nCPU reset.\r\nreadWord(addr=0)=0x20007ffc\r\nreadWord(addr=0xe000ed00)=0x410cc601\r\nreadWord(addr=0x41002018)=0x10010305\r\nwriteWord(addr=0xe000ed0c,value=0x5fa0004)\r\n",
    "stderr": ""
  },
  "version": "1.4.8\r"
}
```

```
$ ./arduino-fwuploader.exe firmware get-version -a COM14 --fqbn arduino:samd:mkrwifi1010 --format json | jq .version
jq: error (at <stdin>:1): Cannot index string with string "version"
"1.4.8\r"
```

Reply:
The first output creates a JSON document of string type:
```json
"Firmware version installed: 1.4.8\r"
```

The second output creates a JSON document of object type:
```json
{
  "programmer": {
    "stdout": "Set binary mode\r\nreadWord(addr=0)=0x20007ffc\r\nreadWord(addr=0xe000ed00)=0x410cc601\r\nreadWord(addr=0x41002018)=0x10010305\r\nversion()=v2.0 [Arduino:XYZ] Apr 11 2019 13:09:56\r\nchipId=0x10010005\r\nConnected at 921600 baud\r\nreadWord(addr=0)=0x20007ffc\r\nreadWord(addr=0xe000ed00)=0x410cc601\r\nreadWord(addr=0x41002018)=0x10010305\r\nAtmel SMART device 0x10010005 found\r\nwrite(addr=0x20004000,size=0x34)\r\nwriteWord(addr=0x20004030,value=0x10)\r\nwriteWord(addr=0x20004020,value=0x20008000)\r\nDevice       : ATSAMD21G18A\r\nreadWord(addr=0)=0x20007ffc\r\nreadWord(addr=0xe000ed00)=0x410cc601\r\nreadWord(addr=0x41002018)=0x10010305\r\nChip ID      : 10010005\r\nversion()=v2.0 [Arduino:XYZ] Apr 11 2019 13:09:56\r\nVersion      : v2.0 [Arduino:XYZ] Apr 11 2019 13:09:56\r\nAddress      : 8192\r\nPages        : 3968\r\nPage Size    : 64 bytes\r\nTotal Size   : 248KB\r\nPlanes       : 1\r\nLock Regions : 16\r\nLocked       : readWord(addr=0x41004020)=0xffff\r\nreadWord(addr=0x41004020)=0xffff\r\nreadWord(addr=0x41004020)=0xffff\r\nreadWord(addr=0x41004020)=0xffff\r\nreadWord(addr=0x41004020)=0xffff\r\nreadWord(addr=0x41004020)=0xffff\r\nreadWord(addr=0x41004020)=0xffff\r\nreadWord(addr=0x41004020)=0xffff\r\nreadWord(addr=0x41004020)=0xffff\r\nreadWord(addr=0x41004020)=0xffff\r\nreadWord(addr=0x41004020)=0xffff\r\nreadWord(addr=0x41004020)=0xffff\r\nreadWord(addr=0x41004020)=0xffff\r\nreadWord(addr=0x41004020)=0xffff\r\nreadWord(addr=0x41004020)=0xffff\r\nreadWord(addr=0x41004020)=0xffff\r\nnone\r\nreadWord(addr=0x41004018)=0\r\nSecurity     : false\r\nBoot Flash   : true\r\nreadWord(addr=0x40000834)=0\r\nBOD          : false\r\nreadWord(addr=0x40000834)=0\r\nBOR          : false\r\nArduino      : FAST_CHIP_ERASE\r\nArduino      : FAST_MULTI_PAGE_WRITE\r\nArduino      : CAN_CHECKSUM_MEMORY_BUFFER\r\nErase flash\r\nchipErase(addr=0x2000)\r\ndone in 0.841 seconds\r\n\r\nWrite 15588 bytes to flash (244 pages)\r\nwrite(addr=0x20005000,size=0x1000)\r\nwriteBuffer(scr_addr=0x20005000, dst_addr=0x2000, size=0x1000)\r\n\r[=======                       ] 26% 
(64/244 pages)write(addr=0x20005000,size=0x1000)\r\nwriteBuffer(scr_addr=0x20005000, dst_addr=0x3000, size=0x1000)\r\n\r[===============               ] 52% (128/244 pages)write(addr=0x20005000,size=0x1000)\r\nwriteBuffer(scr_addr=0x20005000, dst_addr=0x4000, size=0x1000)\r\n\r[=======================       ] 78% (192/244 pages)write(addr=0x20005000,size=0xd00)\r\nwriteBuffer(scr_addr=0x20005000, dst_addr=0x5000, size=0xd00)\r\n\r[==============================] 100% (244/244 pages)\r\ndone in 0.117 seconds\r\n\r\nVerify 15588 bytes of flash with checksum.\r\nchecksumBuffer(start_addr=0x2000, size=0x1000) = e84f\r\nchecksumBuffer(start_addr=0x3000, size=0x1000) = e256\r\nchecksumBuffer(start_addr=0x4000, size=0x1000) = b123\r\nchecksumBuffer(start_addr=0x5000, size=0xce4) = cfbc\r\nVerify successful\r\ndone in 0.016 seconds\r\nCPU reset.\r\nreadWord(addr=0)=0x20007ffc\r\nreadWord(addr=0xe000ed00)=0x410cc601\r\nreadWord(addr=0x41002018)=0x10010305\r\nwriteWord(addr=0xe000ed0c,value=0x5fa0004)\r\n",
    "stderr": ""
  },
  "version": "1.4.8\r"
}
```
The JSON document can't have both a string and object type. This is what causes that `jq` error message.

---
There is also the separate issue of the line break in the `.programmer.stdout` field, which is why GitHub's syntax highlighter put all that red in the output I posted above. It seems to be correctly replacing most of the line breaks with `\r\n` (I guess the `\r` is because this was generated on a Windows machine), but somehow missing that one. line breaks are not permitted in JSON.

Reply:
I have fixed the issue with the printing of the version string even if using `--format json`
It seems I cannot reproduce the issue with the line endings on Linux… I will try to replicate it on Windows.

Reply:
Fixed also the issue with `\r`

I cannot reproduce even on windows.. I tried with PowerShell and the json produced seems to be valid. Maybe you used some other shell that behaves differently?

Reply:
I can no longer reproduce the issue of the line break in the stdout field and the other two issues are resolved. Thanks!